automapper.go

```diff
@@ -47,67 +70,89 @@ func Map(source, dest interface{}) {
+	mapValues(sourceVal, destVal, opts)
 }
 
-func mapValues(sourceVal, destVal reflect.Value, loose bool) {
+func mapValues(sourceVal, destVal reflect.Value, opts MapOptions) {
+	sourceType := sourceVal.Type()
 	destType := destVal.Type()
-	if destType.Kind() == reflect.Struct && sourceVal.Type() != destVal.Type() {
-		if sourceVal.Type().Kind() == reflect.Ptr {
-			if sourceVal.IsNil() {
-				// If source is nil, it maps to an empty struct
-				sourceVal = reflect.New(sourceVal.Type().Elem())
-			}
-			sourceVal = sourceVal.Elem()
-		}
-		for i := 0; i < destVal.NumField(); i++ {
-			mapField(sourceVal, destVal, i, loose)
+	if destType.Kind() == reflect.Struct && sourceVal.Type().Kind() == reflect.Ptr {
+		if sourceVal.IsNil() {
+			sourceVal = reflect.New(sourceType.Elem())
 		}
-	} else if destType == sourceVal.Type() {
+		sourceVal = sourceVal.Elem()
+		mapValues(sourceVal, destVal, opts)
+	} else if destType == sourceType {
 		destVal.Set(sourceVal)
+	} else if destType.Kind() == reflect.Struct && sourceType.Kind() == reflect.Struct {
+		mapFields(sourceVal, destVal, opts)
 	} else if destType.Kind() == reflect.Ptr {
 		if valueIsNil(sourceVal) {
 			return
 		}
 		val := reflect.New(destType.Elem())
-		mapValues(sourceVal, val.Elem(), loose)
+		mapValues(sourceVal, val.Elem(), opts)
 		destVal.Set(val)
 	} else if destType.Kind() == reflect.Slice {
-		mapSlice(sourceVal, destVal, loose)
+		mapSlice(sourceVal, destVal, opts)
 	} else {
 		destVal.Set(sourceVal.Convert(destType))
 	}
 }
 
-func mapSlice(sourceVal, destVal reflect.Value, loose bool) {
+func mapSlice(sourceVal, destVal reflect.Value, opts MapOptions) {
 	destType := destVal.Type()
 	length := sourceVal.Len()
 	target := reflect.MakeSlice(destType, length, length)
 	for j := 0; j < length; j++ {
 		val := reflect.New(destType.Elem()).Elem()
-		mapValues(sourceVal.Index(j), val, loose)
+		mapValues(sourceVal.Index(j), val, opts)
 		target.Index(j).Set(val)
 	}
 
 	if length == 0 {
-		verifyArrayTypesAreCompatible(sourceVal, destVal, loose)
+		verifyArrayTypesAreCompatible(sourceVal, destVal, opts)
 	}
 	destVal.Set(target)
 }
 
-func verifyArrayTypesAreCompatible(sourceVal, destVal reflect.Value, loose bool) {
+func verifyArrayTypesAreCompatible(sourceVal, destVal reflect.Value, opts MapOptions) {
 	dummyDest := reflect.New(reflect.PtrTo(destVal.Type()))
 	dummySource := reflect.MakeSlice(sourceVal.Type(), 1, 1)
-	mapValues(dummySource, dummyDest.Elem(), loose)
+	mapValues(dummySource, dummyDest.Elem(), opts)
+}
+
+func mapFields(sourceVal, destVal reflect.Value, opts MapOptions) {
+	if opts.UseSourceMemberList {
+		for i := 0; i < sourceVal.NumField(); i++ {
+			mapSourceField(sourceVal, destVal, i, opts)
+		}
+	} else {
+		for i := 0; i < destVal.NumField(); i++ {
+			mapDestField(sourceVal, destVal, i, opts)
+		}
+	}
 }
 
-func mapField(source, destVal reflect.Value, i int, loose bool) {
+func mapDestField(source, destVal reflect.Value, i int, opts MapOptions) {
```

Comment:
The signature for `mapDestField` didn't have to be changed. The option `UseSourceMemberList` is not used after mapFields. But I get it if you wanted to have the MapOptions consistently through the whole call stack.

Reply:
`mapDestField` can call `mapValues` which requires the options. And it's nice that we're consistently passing MapOptions, IMO.